src/hooks/transactions/useCheckTransactionStatus/checkBatch.ts

```diff
@@ -0,0 +1,150 @@
+  shouldRefreshBalance?: boolean;
+}
+
+interface RetriesType {
+  [hash: string]: number;
+}
+
+const retries: RetriesType = {};
+const timeouts: string[] = [];
+
+interface ManageTransactionType {
+  serverTransaction: GetTransactionsByHashesReturnType[0];
+  sessionId: string;
+  customTransactionInformation?: CustomTransactionInformation;
+  shouldRefreshBalance?: boolean;
+}
+
+function manageTransaction({
+  serverTransaction,
+  sessionId,
+  customTransactionInformation,
+  shouldRefreshBalance
+}: ManageTransactionType) {
+  const {
+    hash,
+    status,
+    results,
+    invalidTransaction,
+    hasStatusChanged
+  } = serverTransaction;
+  try {
+    if (timeouts.includes(hash)) {
+      return;
+    }
+
+    const retriesForThisHash = retries[hash];
+    if (retriesForThisHash > 30) {
+      // consider transaction as stuck after 1 minute
+      manageTimedOutTransactions(sessionId);
+      return;
+    }
+
+    if (invalidTransaction || getIsTransactionPending(status)) {
+      retries[hash] = retries[hash] ? retries[hash] + 1 : 1;
+      return;
+    }
+    if (hasStatusChanged) {
+      if (
+        getIsTransactionSuccessful(status) &&
+        customTransactionInformation?.completedTransactionsDelay != null
+      ) {
+        //if the transaction is successful and the success status should be updated with a delay
+        //it will enter a timeout and then change the status
+        timeouts.push(hash);
+        setTimeout(
+          () =>
+            store.dispatch(
+              updateSignedTransactionStatus({
+                sessionId,
+                status: TransactionServerStatusesEnum.completed,
+                transactionHash: hash
+              })
+            ),
+          customTransactionInformation?.completedTransactionsDelay
+        );
+      } else {
+        //otherwise, it will just trigger the change of status
+        store.dispatch(
+          updateSignedTransactionStatus({
+            sessionId,
+            status,
+            transactionHash: hash
+          })
+        );
+      }
+    }
+
+    // if set to true will trigger a balance refresh after each iteration
+    if (!shouldRefreshBalance) {
+      refreshAccount();
+    }
+
+    if (getIsTransactionFailed(status)) {
+      manageFailedTransactions({ sessionId, hash, results });
+    }
+  } catch (error) {
+    console.error(error);
+    manageTimedOutTransactions(sessionId);
+  }
+}
+
+export async function checkBatch({
+  sessionId,
+  transactionBatch: { transactions, status, customTransactionInformation },
+  getTransactionsByHash = defaultGetTxByHash,
+  shouldRefreshBalance
+}: TransactionStatusTrackerPropsType) {
+  const isBatchPending = sessionId != null && getIsTransactionPending(status);
+  try {
+    if (!isBatchPending || transactions == null) {
+      return;
+    }
+
+    const pendingTransactions = getPendingTransactions(transactions, timeouts);
+
+    if (pendingTransactions?.length === 0) {
```

Comment:
if the batch is pending, it has pending transactions. Once all transactions go out of pending, only then the batch status changes

Reply:
✅